xtask/src/themelint.rs

```diff
@@ -0,0 +1,156 @@
+use crate::path;
+use crate::DynError;
+use helix_view::theme::Loader;
+use helix_view::theme::Modifier;
+use helix_view::Theme;
+
+struct Rule {
+    fg: Option<&'static str>,
+    bg: Option<&'static str>,
+    check_both: bool,
+}
+
+// Placed in an fn here, so it's the first thing you see
+fn get_rules() -> Vec<Rule> {
+    vec![
+        // Check for ui.selection, which is required
+        Rule::has_either("ui.selection"),
+        // Check for planned readable text
+        Rule::has_fg("ui.text"),
+        Rule::has_bg("ui.background"),
+        // Check for complete editor.statusline bare minimum
+        Rule::has_both("ui.statusline"),
+        Rule::has_both("ui.statusline.inactive"),
+        // Check for editor.color-modes
+        Rule::has_either("ui.statusline.insert"),
+        Rule::has_either("ui.statusline.normal"),
+        Rule::has_either("ui.statusline.select"),
+        // Check for editor.cursorline
+        Rule::has_bg("ui.cursorline.primary"),
+        // Check for editor.rulers
+        Rule::has_either("ui.virtual.ruler"),
+        // Check for menus and prompts
+        Rule::has_both("ui.menu"),
+        Rule::has_both("ui.help"),
+        Rule::has_bg("ui.popup"),
+        Rule::has_either("ui.window"),
+        // Check for visible cursor
+        Rule::has_bg("ui.cursor.primary"),
+        Rule::has_either("ui.cursor.match"),
+    ]
+}
+
+impl Rule {
+    fn has_bg(bg: &'static str) -> Rule {
+        Rule {
+            fg: None,
+            bg: Some(bg),
+            check_both: true,
+        }
+    }
+    fn has_fg(fg: &'static str) -> Rule {
+        Rule {
+            fg: Some(fg),
+            bg: None,
+            check_both: true,
+        }
+    }
+    fn has_either(item: &'static str) -> Rule {
+        Rule {
+            fg: Some(item),
+            bg: Some(item),
+            check_both: false,
+        }
+    }
+    fn has_both(item: &'static str) -> Rule {
+        Rule {
+            fg: Some(item),
+            bg: Some(item),
+            check_both: true,
+        }
+    }
+    fn validate(&self, theme: &Theme, messages: &mut Vec<String>) {
+        let mut found_fg = true;
+        let mut found_bg = true;
+        let mut fg_name = "";
+        let mut bg_name = "";
+        if let Some(fg) = &self.fg {
+            fg_name = fg;
+            if theme.get(fg).fg.is_none() && theme.get(fg).add_modifier == Modifier::empty() {
+                found_fg = false;
+            }
+        }
```

Comment:
I think a better way instead of doing this logic (which seemed weird to me), is to move this part out into a function, same goes for bg, then do
```
if self.check_both {
    if self.found_fg(theme, messages) { ... }
    if self.found_bg(theme, messages) { ... }
} else if !self.found_fg(theme, messages) && !self.found_bg(theme, messages) {
    ...
}
```

Reply:
I've refactored it a bit, but I've kept the handling of messages to `validate` so `found_Xg` only deals with the checking of existence.